File: dreamlike_demo.py
import torch
from transformers import CLIPTextModel, CLIPTokenizer
from diffusers import AutoencoderKL, UNet2DConditionModel, PNDMScheduler
from tqdm.auto import tqdm

vae = AutoencoderKL.from_pretrained(
        "CompVis/stable-diffusion-v1-4",
        subfolder="vae",
        use_safetensors=True
        )

tokenizer = CLIPTokenizer.from_pretrained(
        "dreamlike-art/dreamlike-photoreal-2.0",
        subfolder="tokenizer"
        )

text_encoder = CLIPTextModel.from_pretrained(
        "dreamlike-art/dreamlike-photoreal-2.0",
        subfolder="text_encoder",
        use_safetensors=True
        )

unet = UNet2DConditionModel.from_pretrained(
        "dreamlike-art/dreamlike-photoreal-2.0",
        subfolder="unet",
        use_safetensors=True
        )

from diffusers import UniPCMultistepScheduler

scheduler = UniPCMultistepScheduler.from_pretrained(
        "dreamlike-art/dreamlike-photoreal-2.0",
        subfolder="scheduler"
        )

torch_device = "cpu" # "cuda"
vae.to(torch_device)
text_encoder.to(torch_device)
unet.to(torch_device)

# from diffusers import AutoPipelineForText2Image
#
# pipe_txt2img = AutoPipelineForText2Image.from_pretrained(
#     "dreamlike-art/dreamlike-photoreal-2.0", torch_dtype=torch.float16, use_safetensors=True
# ).to("cuda")
#
# generator = torch.Generator(device="cuda").manual_seed(37)
# image = pipe_txt2img(prompt, generator=generator).images[0]
# image

prompt = ["godzilla is watching kitty doing homework"]
height = 512  # default height of Stable Diffusion
width = 512  # default width of Stable Diffusion
num_inference_steps = 25  # Number of denoising steps
guidance_scale = 7.5  # Scale for classifier-free guidance
generator = torch.manual_seed(0)  # Seed generator to create the initial latent noise
batch_size = len(prompt)

text_input = tokenizer(
        prompt,
        padding="max_length",
        max_length=tokenizer.model_max_length,
        truncation=True,
        return_tensors="pt"
        )
print(text_input.input_ids)

with torch.no_grad():
    text_embeddings = text_encoder(text_input.input_ids.to(torch_device))[0]

print(text_embeddings)

latents = torch.randn(
        (batch_size, unet.config.in_channels, height // 8, width // 8),
        generator=generator,
        device=torch_device,
        )
latents = latents * scheduler.init_noise_sigma


scheduler.set_timesteps(num_inference_steps)


for t in tqdm(scheduler.timesteps):
    # expand the latents if we are doing classifier-free guidance to avoid doing two forward passes.
    latent_model_input = torch.cat([latents] * 2)

    latent_model_input = scheduler.scale_model_input(latent_model_input, timestep=t)

    # predict the noise residual
    with torch.no_grad():
        noise_pred = unet(latent_model_input, t, encoder_hidden_states=text_embeddings).sample

    breakpoint()
    # perform guidance
    noise_pred_uncond, noise_pred_text = noise_pred.chunk(2)
    noise_pred = noise_pred_uncond + guidance_scale * (noise_pred_text - noise_pred_uncond)

    # compute the previous noisy sample x_t -> x_t-1
    latents = scheduler.step(noise_pred, t, latents).prev_sample

File: dreamlike_demo.rb
require "torch"
require "onnxruntime"
require "tokenizers"
require "debug"

providers = if Torch::CUDA.available?
              %w[CUDAExecutionProvider]
            else
              %w[CPUExecutionProvider]
            end
device = "cpu" # use "cuda" with gpu, currently use cpu temporarily to handle gpu runtime error

# model:
# https://huggingface.co/docs/diffusers/tutorials/autopipeline
#
# run steps:
# https://huggingface.co/docs/diffusers/using-diffusers/write_own_pipeline#deconstruct-the-stable-diffusion-pipeline
#
# scheduler: there is no Ruby implementation yet.

# Create text tokens
prompt = ["godzilla is watching kitty doing homework"]
batch_size = prompt.length

# Create text tokens
tokenizer = Tokenizers.from_pretrained("openai/clip-vit-base-patch32")
tokenizer.enable_padding(length: 77, pad_id: 49407)
tokenizer.enable_truncation(77)
text_tokens = tokenizer.encode_batch(prompt)
text_ids = Torch.tensor(text_tokens.map(&:ids))
pp text_ids

# Create text embeddings
text_encoder = OnnxRuntime::Model.new("../onnx/text_encoder/model.onnx", providers:) # name: openai/clip-vit-large-patch14
text_embeddings = Torch.no_grad do
  text_encoder
    .predict({ input_ids: text_ids }) # Shape: 1x77
    .then { |h| Torch.tensor(h["last_hidden_state"]) } # Shape: 1x77x768
end
pp text_embeddings

# Generate the unconditional text embeddings which are the embeddings for the padding token.
# max_length = text_ids.shape[-1] # 77
uncond_tokens = tokenizer.encode_batch([""] * batch_size)
uncond_ids = Torch.tensor(uncond_tokens.map(&:ids))
uncond_embeddings = text_encoder.predict({ input_ids: uncond_ids })
  .then { |h| Torch.tensor(h["last_hidden_state"]) } # Shape: 1x77x768

text_embeddings = Torch.cat([uncond_embeddings, text_embeddings])

# Create random noise
height = 512
width = 512
unet = OnnxRuntime::Model.new("../onnx/unet/model.onnx", providers:)
channels_num = unet.inputs.detect{ |e| e[:name] == "sample" }[:shape][1]
generator = Torch::Generator.new.manual_seed(0) # Seed generator to create the initial latent noise
latents = Torch.randn([batch_size, channels_num, height / 8, width / 8], generator:, device:) # Shape: 1x4x64x64

# vae_encoder = OnnxRuntime::Model.new("../onnx/vae_encoder/model.onnx", providers:)
# vae_decoder = OnnxRuntime::Model.new("./onnx/vae_decoder/model.onnx", providers:)
#

num_inference_steps = 25 # denoising steps
guidance_scale = 7.5 # classifier-free guidance

debugger
